[libc++][functional] Implement `std::bind_front<NTTP>`

JMazurkiewicz opened the pull request:
Add `std::bind_front<NTTP>` function from [P2714R1](https://wg21.link/p2714r1).

Towards https://github.com/llvm/llvm-project/issues/105388.

JMazurkiewicz committed:
[libc++][functional] Implement `bind_front<NTTP>`

JMazurkiewicz committed:
[libc++][functional] Test `bind_front<NTTP>`

frederick-vs-ja commented on libcxx/include/__functional/bind_front.h:
```diff
@@ -49,6 +51,38 @@ _LIBCPP_HIDE_FROM_ABI constexpr auto bind_front(_Fn&& __f, _Args&&... __args) {
+
+template <auto _Fn, class... _BoundArgs>
+struct __nttp_bind_front_t : __perfect_forward<__nttp_bind_front_op<_Fn>, _BoundArgs...> {
+  using __perfect_forward<__nttp_bind_front_op<_Fn>, _BoundArgs...>::__perfect_forward;
```
We should be able to use explicit `this` parameters.

JMazurkiewicz committed:
Remove unnecessary `{}`

JMazurkiewicz committed:
Fix `bind_front.nttp.verify.cpp` test

JMazurkiewicz committed:
Use explicit `this` parameter instead of `__perfect_forward` base class

JMazurkiewicz committed:
Return different* function object when `sizeof...(_BoundArgs) == 0`

* Such that `std::is_empty_v<F>` is `true` and `operator()` is static.

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/bind_front_nttp

huixie90 commented on libcxx/include/__functional/bind_front.h:
```diff
@@ -49,6 +51,53 @@ _LIBCPP_HIDE_FROM_ABI constexpr auto bind_front(_Fn&& __f, _Args&&... __args) {
+  static_assert((is_move_constructible_v<decay_t<_Args>> && ...),
+                "bind_front requires all decay_t<Args> to be move constructible");
+  if constexpr (using _Ty = decltype(_Fn); is_pointer_v<_Ty> || is_member_pointer_v<_Ty>)
+    static_assert(_Fn != nullptr, "f cannot be equal to nullptr");
```
Could we have a verify test where we pass a nullptr and this static assertion triigered?

huixie90 commented on libcxx/include/__functional/bind_front.h:
```diff
@@ -49,6 +51,53 @@ _LIBCPP_HIDE_FROM_ABI constexpr auto bind_front(_Fn&& __f, _Args&&... __args) {
+
+template <auto _Fn, class... _Args>
+[[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto bind_front(_Args&&... __args) {
+  static_assert((is_constructible_v<decay_t<_Args>, _Args> && ...),
```
can we also test these `static_assert` in a verify test?

JMazurkiewicz replied:
Already here: https://github.com/JMazurkiewicz/llvm-project/blob/libcxx/bind_front_nttp/libcxx/test/std/utilities/function.objects/func.bind.partial/bind_front.nttp.verify.cpp#L28-L51

huixie90 commented on libcxx/test/std/utilities/function.objects/func.bind.partial/bind_front.nttp.pass.cpp:
```diff
@@ -0,0 +1,355 @@
+    }
+  }
+
+  { // Test perfect forwarding
```
I think in the test suite what is missing and important is that the bond arguments are correct perfectly forward-like  as the returned object. given
```
constexpr struct F {
 
  int operator()(int&, double) const;
  int operator()(const int&, double) const;
  int operator()(int&&, double) const;
  int operator()(const int&&, double) const;
} f{};
```
and

```
auto r = bind_front<f>(5);
```

the results of
```
r(1.0);
std::as_const(r)(1.0);
std::move(r)(1.0);
std::move(std::as_const(r))(1.0);
```

should call different overloads

JMazurkiewicz replied:
Done: https://github.com/llvm/llvm-project/pull/165096/commits/4689f86dfe437a356fd296328783ed5c9c276fe6

JMazurkiewicz committed:
Mention `bind_front` in `static_assert(_Fn != nullptr)` message

JMazurkiewicz committed:
Add better test of perfect forwarding of bound arguments

JMazurkiewicz committed:
Test exception propagation

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/bind_front_nttp

JMazurkiewicz committed:
Fix `bind_front.nttp.verify.cpp` test

huixie90 reviewed (commented):
Thanks a lot ! It looks good to me once my new comments are addressed and CI is green. I will leave final approval to Nicolas as he also has some open comments

JMazurkiewicz committed:
Just `static_assert(test())`

JMazurkiewicz committed:
Improve perfect forwarding test a little bit

Comment: https://github.com/llvm/llvm-project/pull/165096#discussion_r2553936718

>the test looks nice. but it would be even better if we can let the test to run the function instead of just check the type

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/bind_front_nttp

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/bind_front_nttp

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/bind_front_nttp

frederick-vs-ja reviewed (approved):
LGTM. I'd like to merge this soon if there's no more change request.

huixie90 reviewed (approved):
Sorry for the delay

frederick-vs-ja commented:
I removed the sentence mentioning the previous PR, because I don't think it's suitable for the squashed commit message. Please double-check this.

frederick-vs-ja commented:
Oh, we seemed to forget bumping `__cpp_lib_bind_front`.

JMazurkiewicz commented:
> Oh, we seemed to forget bumping `__cpp_lib_bind_front`.

It was bumped (incorrectly) in 3f05d044f491b97596fdaab7614f81cfb70ba54d.
